### PR TITLE
instantsend: Remove islocks for rejected/removed txes

### PR DESCRIPTION
**Pull request by UdjinM6**
~Based on https://github.com/dashpay/dash/pull/4154 atm~

### CODE REVIEW

**Review by PastaPastaPasta — DISMISSED** [review]
utACK

**Review by xdustinface — DISMISSED** [review]
utACK but wouldn't it make sense to add some `TODO: Remove later / with version x` comment for `CInstantSendDb::Upgrade` and maybe add a DB version define instead of the `1`?

**Comment by UdjinM6** [comment]
> wouldn't it make sense to add some `TODO: Remove later / with version x` comment for `CInstantSendDb::Upgrade`

Not necessarily. You could implement more upgrades later if needed. The actual upgrade is only done once and `Read` is cheap. I think it's fine to keep it.

> maybe add a DB version define instead of the `1`

Good point 👍 see 94ec87a

**Review by xdustinface — APPROVED** [review]
utACK

**Comment by UdjinM6** [comment]
Mac build was not happy because of a missing definition... force-pushed

**Review by xdustinface — APPROVED** [review]
utACK

**Comment by UdjinM6** [comment]
I _believe_ https://github.com/dashpay/dash/pull/4155/commits/d9d8c0ffcd12fd12a1cecc879e2e237e79a51478 fixed issues I had on one of nodes I run to test this. At least I no longer see unexpected `TransactionRemovedFromMempool` in `debug.log` and tx count vs islock count looks sane now.

**Review by PastaPastaPasta — APPROVED** [review]
utACK, let's get this on testnet

File: src/llmq/quorums_instantsend.cpp [hunk 1]
```diff
@@ -1026,6 +1070,23 @@ void CInstantSendManager::TransactionAddedToMempool(const CTransactionRef& tx)
     }
 }
 
+void CInstantSendManager::TransactionRemovedFromMempool(const CTransactionRef& tx)
+{
+    if (!IsInstantSendEnabled() || !masternodeSync.IsBlockchainSynced() || tx->vin.empty()) {
```
Comment (xdustinface) [inline src/llmq/quorums_instantsend.cpp:1075]:
```suggestion
    if (tx->vin.empty()) {
```
Shouldn't it always at least try to remove it?

**Review by PastaPastaPasta — APPROVED** [review]
re-utACK

**Review by xdustinface — APPROVED** [review]
utACK

**Pull request merged**